### PR TITLE
Fix DialContext when using a timeout

### LINKED ISSUE #2736: ClientConn: WithTimeout() causes Dial error to be always Canceled
A solution is rearrange defer functions in DialContext().

https://github.com/grpc/grpc-go/pull/2733#discussion_r271705181

### PR DESCRIPTION
**Pull request by daviddrysdale**
Fixes #2736

### CODE REVIEW

**Comment by thelinuxfoundation** [comment]
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* User @daviddrysdale isn't covered by a CLA. They will need to complete the form at https://identity.linuxfoundation.org/projects/cncf

_Regards,
CLA GitHub bot_

**Comment by daviddrysdale** [comment]
CNCF account created as a Google employee.

File: clientconn_test.go [hunk 1]
```diff
@@ -51,6 +51,48 @@ func assertState(wantState connectivity.State, cc *ClientConn) (connectivity.Sta
...
+		}
+		<-dialDone // Close conn only after dial returns.
+	}()
+
+	r, cleanup := manual.GenerateAndRegisterManualResolver()
+	defer cleanup()
+	r.InitialState(resolver.State{Addresses: []resolver.Address{lisAddr}})
+	client, err := Dial(r.Scheme()+":///test.server", WithInsecure(), WithTimeout(5*time.Second))
+	close(dialDone)
+	if err != nil {
```
Comment (menghanl) [inline clientconn_test.go]: Are you still working on this test? It doesn't seem to test the error behavior.

Comment (daviddrysdale) [reply clientconn_test.go]: Er, not really :-).   I got it just far enough so it fails with the `DialContext` timeout, but I don't know what else would be worth testing after that.

Comment (menghanl) [reply clientconn_test.go]: Oh, now I see what it's testing.

**Pull request merged**